File: src/flame/net/unix_server.cpp
#include "deps.h"
#include "../flame.h"
#include "../coroutine.h"
#include "net.h"
#include "unix_socket.h"
#include "unix_server.h"

namespace flame {
namespace net {
	unix_server::unix_server()
	: svr((uv_pipe_t*)malloc(sizeof(uv_pipe_t)))
	, acc((uv_stream_t*)svr)
	, bound_(false) {
		uv_pipe_init(flame::loop, svr, 0);
	}
	unix_server::~unix_server() {
		close();
	}
	php::value unix_server::handle(php::parameters& params) {
		acc.handle(params[0]);
		return nullptr;
	}
	php::value unix_server::bind(php::parameters& params) {
		if(bound_) throw php::exception("failed to bind: server is already bound");
		
		php::string path = params[0].to_string();
		if(path.c_str()[0] != '/') {
			throw php::exception("failed to bind: only absolute path is allowed");
		}
	
		int err = uv_pipe_bind(svr, path.c_str());
		if(err < 0) {
			throw php::exception(uv_strerror(err), err);
		}
		bound_ = true;
		return nullptr;
	}
	php::value unix_server::run(php::parameters& params) {
		if(!bound_) throw php::exception("failed to run: server is not bound");
		
		acc.run();
		return flame::async(this);
	}
	php::value unix_server::close(php::parameters& params) {
		close();
		return nullptr;
	}
	void unix_server::close() {
		if(svr) {
			acc.close();
			uv_close((uv_handle_t*)svr, flame::free_handle_cb);
			svr = nullptr;
		}
	}
}
}

File: test/flame/net/unix_server.php
<?php
flame\init("unix_server_test");
flame\go(function() {
	$server = new flame\net\unix_server();
	// 每个连接会回调到单独的协程函数中进行处理
	$server->handle(function($sock) {
		var_dump($sock);
		$i = 0;
		while($i<1000) {
			try{
				yield $sock->write("".(++$i));
				yield flame\time\sleep(100);
			}catch(exception $ex) {
				// 客户端异常关闭连接会产生 BROKEN_PIPE 等错误需要处理
				var_dump($ex);
				break;
			}
		}
	});
	@unlink("/data/sockets/flame.xingyan.panda.tv.sock");
	$server->bind("/data/sockets/flame.xingyan.panda.tv.sock");
	// 在 $server->close() 之前，会“阻塞”在此处
	yield $server->run();
	echo "======\n";
});
flame\run();
